### PR TITLE
[eas-cli] support multiflavor android projects

### PR DESCRIPTION
**Pull request by wkozyra95**
blocked by https://github.com/expo/expo-cli/pull/3830

# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.
- [ ] I've tagged the changelog entry with `[EAS BUILD API]` if it's a part of a breaking change to EAS Build API (only possible when updating `@expo/eas-build-job` package).

# Why

applicationId and versions are read incorrectly if there are multiple flavors defined
this PR adds support for
- build command
- credentials command

# How

- [moved gradle utils from expo-cli to eas-cli] https://github.com/expo/expo-cli/pull/3830
- if parsing fails fallback to old behavior for applicationId

# Test Plan

fixed existing tests
tested by running builds and credentials manager on a test project

### CODE REVIEW

File: packages/eas-cli/src/build/ios/version.ts [hunk 1]
```diff
@@ -136,6 +136,20 @@ export async function readBuildNumberAsync(
...
+  exp: ExpoConfig,
+  buildSettings: XCBuildConfiguration['buildSettings']
+): Promise<{ appVersion?: string; appBuildVersion?: string }> {
+  try {
+    return {
+      appBuildVersion: await readBuildNumberAsync(projectDir, exp, buildSettings),
+      appVersion: await readShortVersionAsync(projectDir, exp, buildSettings),
+    };
+  } catch {}
+  return {};
```
Comment (dsokal) [inline packages/eas-cli/src/build/ios/version.ts:150]:
```suggestion
  } catch {
    return {};
  }
```

File: packages/eas-cli/src/build/metadata.ts [hunk 2]
```diff
@@ -69,38 +74,38 @@ export async function collectMetadata<T extends Platform>(
...
-  if (ctx.platform === Platform.IOS) {
-    const iosContext = platformContext as IosMetadataContext | undefined;
-    return await readShortVersionAsync(ctx.projectDir, ctx.exp, iosContext?.buildSettings ?? {});
-  } else {
-    return await readVersionNameAsync(ctx.projectDir, ctx.exp);
-  }
-}
-
-async function resolveAppBuildVersionAsync<T extends Platform>(
+async function resolveVersionsAsync<T extends Platform>(
```
Comment (dsokal) [inline packages/eas-cli/src/build/metadata.ts:77]:
```suggestion
async function maybeResolveVersionsAsync<T extends Platform>(
```

File: packages/eas-cli/src/build/metadata.ts [hunk 3]
```diff
@@ -69,38 +74,38 @@ export async function collectMetadata<T extends Platform>(
...
   ctx: BuildContext<T>,
   platformContext?: MetadataContext<T>
```
Comment (dsokal) [inline packages/eas-cli/src/build/metadata.ts:79]:
```suggestion
  platformContext: MetadataContext<T>
```
It seems it's always defined.

File: packages/eas-cli/src/build/metadata.ts [hunk 4]
```diff
@@ -69,38 +74,38 @@ export async function collectMetadata<T extends Platform>(
...
-): Promise<string | undefined> {
+): Promise<{ appBuildVersion?: string; appVersion?: string }> {
   if (ctx.platform === Platform.IOS) {
     const iosContext = platformContext as IosMetadataContext | undefined;
```
Comment (dsokal) [inline packages/eas-cli/src/build/metadata.ts:82]:
```suggestion
    const iosContext = platformContext as IosMetadataContext;
```

File: packages/eas-cli/src/credentials/android/AndroidCredentialsProvider.ts [hunk 5]
```diff
@@ -34,7 +34,7 @@ export default class AndroidCredentialsProvider {
   }
 
   private async getRemoteAsync(): Promise<AndroidCredentials> {
-    const setupBuildCredentialsAction = await new SetupBuildCredentials({ app: this.options.app });
+    const setupBuildCredentialsAction = new SetupBuildCredentials({ app: this.options.app });
```
Comment (dsokal) [inline packages/eas-cli/src/credentials/android/AndroidCredentialsProvider.ts:37]: 👍

File: packages/eas-cli/src/project/android/applicationId.ts [hunk 6]
```diff
@@ -22,28 +24,44 @@ export async function ensureApplicationIdIsDefinedForManagedProjectAsync(
   assert(workflow === Workflow.MANAGED, 'This function should be called only for managed projects');
 
   try {
-    return await getApplicationIdAsync(projectDir, exp);
+    return await getApplicationIdAsync(projectDir, exp, { moduleName: 'app' });
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/applicationId.ts:27]:
```suggestion
    return await getApplicationIdAsync(projectDir, exp, { moduleName: DEFAULT_MODULE_NAME });
```
or sth

File: packages/eas-cli/src/project/android/applicationId.ts [hunk 7]
```diff
@@ -22,28 +24,44 @@ export async function ensureApplicationIdIsDefinedForManagedProjectAsync(
...
   } catch (err) {
     return await configureApplicationIdAsync(projectDir, exp);
   }
 }
 
-export async function getApplicationIdAsync(projectDir: string, exp: ExpoConfig): Promise<string> {
+export async function getApplicationIdAsync(
+  projectDir: string,
+  exp: ExpoConfig,
+  gradleContext: GradleBuildContext | undefined
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/applicationId.ts:36]:
```suggestion
  gradleContext?: GradleBuildContext
```

File: packages/eas-cli/src/project/android/applicationId.ts [hunk 8]
```diff
@@ -22,28 +24,44 @@ export async function ensureApplicationIdIsDefinedForManagedProjectAsync(
...
-    } catch {}
-    if (!buildGradlePath || !fs.pathExistsSync(buildGradlePath)) {
-      throw new Error(errorMessage);
+    if (gradleContext) {
+      const buildGradle = await getAppBuildGradleAsync(projectDir);
+      const applicationId = resolveConfigValue(buildGradle, 'applicationId', gradleContext.flavor);
+      return nullthrows(applicationId, errorMessage);
+    } else {
+      // fallback to best effort approach, this logic can be dropped when we start supporting
+      // modules diffrent than 'app' and 'flavorDimensions'
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/applicationId.ts:49]:
```suggestion
      // modules different than 'app', and 'flavorDimensions'
```

File: packages/eas-cli/src/build/android/version.ts [hunk 9]
```diff
@@ -1,50 +1,41 @@
...
-    return exp.version;
-  }
-}
-
-function readBuildGradle(projectDir: string): string | undefined {
-  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
-  if (!fs.pathExistsSync(buildGradlePath)) {
-    return undefined;
+    return {
+      appBuildVersion: String(AndroidConfig.Version.getVersionCode(exp) ?? 1),
```
Comment (dsokal) [inline packages/eas-cli/src/build/android/version.ts:37]:
```suggestion
      appBuildVersion: String(AndroidConfig.Version.getVersionCode(exp)),
```
`getVersionCode` never returns falsy values. 
https://github.com/expo/expo-cli/commit/e3dab0c9fc917c490cf086c0cbe0fbdb950dbc33

File: packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts [hunk 10]
```diff
@@ -60,7 +60,7 @@ export default class AndroidSubmitCommand {
 
   private async maybeGetAndroidPackageFromCurrentProjectAsync(): Promise<string | undefined> {
     try {
-      return await getApplicationIdAsync(this.ctx.projectDir, this.ctx.exp);
+      return await getApplicationIdAsync(this.ctx.projectDir, this.ctx.exp, { moduleName: 'app' });
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts:63]: Are you planning to change this in a follow-up PR?

File: packages/eas-cli/src/project/android/gradleUtils.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
+  const rawBuildGradle = await fs.readFile(buildGradlePath, 'utf8');
+  return await g2js.parseText(rawBuildGradle);
+}
+
+export function resolveConfigValue(
+  buildGradle: AppBuildGradle,
+  field: keyof Config,
+  flavor?: string
+) {
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/gradleUtils.ts:37]: Missing return type.

Comment (dsokal) [reply packages/eas-cli/src/project/android/gradleUtils.ts:37]: Can you check if it's easily possible (without too many changes) to turn on the eslint rule to require function return types?

File: packages/eas-cli/src/project/android/gradleUtils.ts [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+  const rawCmd = withoutParams.startsWith(':') ? withoutParams.slice(1) : withoutParams;
+
+  // separate moduleName and rest of the definition
+  const splitCmd = rawCmd.split(':');
+  const [moduleName, taskName] =
+    splitCmd.length > 1 ? [splitCmd[0], splitCmd[1]] : [undefined, splitCmd[0]];
+
+  const matchResult = taskName.match(/(build|bundle|assemble)(.*)(Release|Debug)/);
+  if (!matchResult) {
+    throw new Error('Failed to parse gradle command');
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/gradleUtils.ts:76]:
```suggestion
    throw new Error('Failed to parse gradle command: ${...}');
```

File: packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts (hunk 10 above)
Comment (wkozyra95) [reply packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts:63]: yes

File: packages/eas-cli/src/project/android/gradle.ts [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+): Promise<GradleBuildContext | undefined> {
+  const workflow = await resolveWorkflowAsync(projectDir, Platform.ANDROID);
+  if (workflow === Workflow.GENERIC) {
+    try {
+      if (buildProfile.gradleCommand) {
+        const buildGradle = await gradleUtils.getAppBuildGradleAsync(projectDir);
+        const parsedGradleCommand = buildProfile.gradleCommand
+          ? gradleUtils.parseGradleCommand(buildProfile.gradleCommand, buildGradle)
+          : undefined;
+        if (parsedGradleCommand?.moduleName && parsedGradleCommand.moduleName !== 'app') {
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/gradle.ts:25]:
```suggestion
        if (parsedGradleCommand?.moduleName && parsedGradleCommand.moduleName !== DEFAULT_MODULE_NAME) {
```

File: packages/eas-cli/src/project/android/gradle.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+          Log.warn('Building modules different than "app" migth result in unexpected behavior');
```
Comment (dsokal) [inline packages/eas-cli/src/project/android/gradle.ts:26]:
```suggestion
          Log.warn(`Building modules different than "${DEFAULT_MODULE_NAME}" might result in unexpected behavior`);
```

Comment (dsokal) [reply packages/eas-cli/src/project/android/gradle.ts:26]: Note the typo! `migth` -> `might`

**Review by dsokal — APPROVED** [review]

File: packages/eas-cli/src/project/android/gradleUtils.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+export const DEFAULT_MODULE_NAME = 'app';
+
+export async function getAppBuildGradleAsync(projectDir: string): Promise<AppBuildGradle> {
+  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
+  const rawBuildGradle = await fs.readFile(buildGradlePath, 'utf8');
+  return await g2js.parseText(rawBuildGradle);
+}
+
+export function resolveConfigValue(
```
Comment (jkhales) [inline packages/eas-cli/src/project/android/gradleUtils.ts:35]: How are you planning to support writing to multiflavor gradles?

e.g. what is the analog to this? https://github.com/expo/expo-cli/blob/master/packages/config-plugins/src/android/Version.ts#L40

In general is multiflavor support only intended for EAS projects. Why are these gradle utils are in `eas-cli` intead of `config-plugins`?

Comment (wkozyra95) [reply packages/eas-cli/src/project/android/gradleUtils.ts:35]: >  Why are these gradle utils are in eas-cli intead of config-plugins?

https://github.com/expo/expo-cli/pull/3830/files#r704836238

> How are you planning to support writing to multiflavor gradles?

I don't think we should, we should just limit autoincrement functionality for managed and simple bare projects(by simple I mean no favors on android and one scheme/target on iOS). With projects like that, we are likely to cause more problems than the feature is worth.

Comment (jkhales) [reply packages/eas-cli/src/project/android/gradleUtils.ts:35]: OK cool, seems like we are on the same page with both issues. 

Will continue monitoring discussion on your slack post.

**Pull request merged**